### PR TITLE
Pr move velocity constraints to end

### PR DESCRIPTION
**Pull request by Stifael**
this is a quick fix for the current spike. basically no matter what velocity or thrust computation has been done, the setpoints need to be constraint at the end. 
https://github.com/PX4/Firmware/issues/7535

### CODE REVIEW

**Comment by Stifael** [comment]
there are many other fixes in terms of auto in this PR (several hours of flight tests), which will solves many of the current issues that we experience. I just need to spend some time to finish the integration

**Comment by LorenzMeier** [comment]
@PX4TestFlights please test as soon as feasible. Thanks!

**Comment by santiago3dr** [comment]
couple of flight running  pixracer (V4)

overall good flights; no issues
http://logs.px4.io/plot_app?log=2b762d8d-606e-4273-b5a9-76d873c3ba7f
http://logs.px4.io/plot_app?log=a39851a0-9873-4465-b17f-e9dd424bf0ae

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -1694,14 +1709,14 @@ MulticopterPositionControl::calculate_velocity_setpoint(float dt)
...
-		_vel_sp(2) = (_pos_sp(2) - _pos(2)) * _params.pos_p(2);
-	}
-
-	/* make sure velocity setpoint is saturated in xy*/
-	float vel_norm_xy = sqrtf(_vel_sp(0) * _vel_sp(0) +
-				  _vel_sp(1) * _vel_sp(1));
+		if (PX4_ISFINITE(_pos_sp(2))) {
+			_vel_sp(2) = (_pos_sp(2) - _pos(2)) * _params.pos_p(2);
 
-	slow_land_gradual_velocity_limit();
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: You just kicked this out without replacement. I created a nicer replacement of this but it's not integrated yet.

**Comment by MaEtUgR** [comment]
rebased on master without conflicts

**Comment by MaEtUgR** [comment]
I took the time to go through all the enhancements we did relating to velocity setpoint generation and limiting and integrated them carefully.

Here's a extract of the commit message explained further:
- one warn_rate_limited was missing
(only output, no functionality)
- vel_sp_slewrate was in the wrong order for smooth slowdown/speedup on takeoff and landing
(order of limiting different)
- slow_land_gradual_velocity_limit was replaced by calls to math::gradual
(this is why I started to interfere see: https://github.com/PX4/Firmware/pull/7538#discussion_r126058063)
- smooth takeoff speed got controllable by user input
(very small enhancement, not really related)
- comments were corrected
(no functionality but a lot clearer to read)
- an additional check for the sanity of velocity setpoints was added
(helps with general setpoint generation problems)

@Stifael sorry, you can kill me later

**Comment by LorenzMeier** [comment]
@Stifael Can you fly this on the F450 DIY frame?

**Comment by LorenzMeier** [comment]
@PX4TestFlights Could you please fly this state again?

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 2]
```diff
@@ -1727,36 +1719,49 @@ MulticopterPositionControl::calculate_velocity_setpoint(float dt)
...
-
-		_vel_sp(2) = math::max(_vel_sp(2), -_params.tko_speed);
+		float vel_limit = math::gradual(altitude_above_home,
+						_params.slow_land_alt2, _params.slow_land_alt1,
+						_params.tko_speed, _params.vel_max_up);
+		_vel_sp(2) = math::max(_vel_sp(2), -vel_limit);
 	}
 
+	/* limit vertical downwards speed (positive z) close to ground
+	 * for now we use the altitude above home and assume that we want to land at same hight as we took off */
```
Comment (Stifael) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: height

**Comment by Stifael** [comment]
@LorenzMeier yep

**Comment by Stifael** [comment]
unfortunately I was not able to test it yet. I had problems to update parameters and it also did not allow me to do rtl over mavlink (did not even get any rejection messages). @PX4TestFlights please can you pay attention what happens when you switch into rtl (in particular from position control or mission)?

**Comment by LorenzMeier** [comment]
Did you arm before you had home set?

**Comment by Stifael** [comment]
@LorenzMeier not sure, but it was not the only problem I had. 
I also tried to test it with the aero, but unfortunately the testing did not go very well. it overshot the waypoints a lot. Even though I don't think that had to do anything with this pr, it just means that I cannot conclude anything (the logging was also turned off because that is the default for some reason)

**Comment by Avysuarez** [comment]
Couple of flights with Pixhawk mini (V3). Good flights no issues.
http://review.px4.io/plot_app?log=550143e6-f3bb-4ceb-aa05-286107a3beeb
http://review.px4.io/plot_app?log=67fb9daf-95a4-4839-969b-06606cb5153f

**Comment by BazookaJoe1900** [comment]
Does this fix should also fix what happen in 'Flyaway after loss of datalink (v1.6.2) #7535'?
note that the heading was wrong in that scenario.
Also, it seems like critical fix, is there going to be 1.6.4 version with this fix?

**Comment by MaEtUgR** [comment]
@BazookaJoe1900 Independent of any possible additional bug that might cause the vehicle to fly into the wrong direction this PR should always assure that the velocity setpoint cannot under any circumstances exceed the maximum configured by parameter `MPC_XY_VEL_MAX`. This means it should also fix the extreme setpoint spikes to 430m/s from https://github.com/PX4/Firmware/issues/7535 but might not fix the root cause of the issue. Feedback whether it's fixed or not would be appreciated, it should now be safer to test because the setpoint is usefully bounded again.

**Pull request merged**